Sources/SwiftProtobuf/AnyMessageStorage.swift

```diff
@@ -301,163 +323,173 @@ extension AnyMessageStorage {
+        //    orders, so the messages could be identical, but in binary form
+        //    they could differ.
+        public func hash(into hasher: inout Hasher) {
+          if !_typeURL.isEmpty {
+            hasher.combine(_typeURL)
+          }
+        }
+      #else  // swift(>=4.2)
+        var hashValue: Int {
+          var hash: Int = i_2166136261
+          if !_typeURL.isEmpty {
+            hash = (hash &* i_16777619) ^ _typeURL.hashValue
+          }
+          return hash
+        }
+      #endif
+
+      func isEqualTo(other: AnyMessageStorage) -> Bool {
+            if _typeURL != other._typeURL { return false }
+
+            // Since the library does lazy Any decode, equality is a very hard problem.
+            // It things exactly match, that's pretty easy, otherwise, one ends up having
+            // to error on saying they aren't equal.
+            //
+            // The best option would be to have Message forms and compare those, as that
+            // removes issues like map<> serialization order, some other protocol buffer
+            // implementation details/bugs around serialized form order, etc.; but that
+            // would also greatly slow down equality tests.
+            //
+            // Do our best to compare what is present have...
+
+            // If both have messages, check if they are the same.
+            if case .message(let myMsg) = state,
+                  case .message(let otherMsg) = other.state, type(of: myMsg)
+                  == type(of: otherMsg) {
+                  // Since the messages are known to be same type, we can claim both equal and
+                  // not equal based on the equality comparison.
+                  return myMsg.isEqualTo(message: otherMsg)
+            }
+
+            // If both have serialized data, and they exactly match; the messages are equal.
+            // Because there could be map in the message, the fact that the data isn't the
+            // same doesn't always mean the messages aren't equal. Likewise, the binary could
+            // have been created by a library that doesn't order the fields, or the binary was
+            // created using the appending ability in of the binary format.
+            if case .binary(let myValue) = state,
+                  case .binary(let otherValue) = other.state, myValue
+                  == otherValue
+            { return true }
+
+            // If both have contentJSON, and they exactly match; the messages are equal.
+            // Because there could be map in the message (or the JSON could just be in a different
+            // order), the fact that the JSON isn't the same doesn't always mean the messages
+            // aren't equal.
+            if case .contentJSON(let myJSON, _) = state,
+                  case .contentJSON(let otherJSON, _) = other.state, myJSON
+                  == otherJSON
+            { return true }
```

Comment:
Do we hate this? (Allowing the body to be on a single line below a prologue that wrapped?)

I'm ambivalent; it's not terrible, and it would take a bit more work to make node wrapping dependent on other nodes. If we were respecting discretionary newlines, of course, the user could put them in and they'd stick.

Reply:
I don't hate it enough to complicate the rules, I think.  With discretionary breaks most people will do the palatable thing and there won't be a problem.  Now the interesting thing is, if we are respecting discretionary breaks but they give us this nasty thing

```swift
if long
    + condition {
    return true }
```

Will we insert the newline before the close brace as follows

```swift
if long
    + condition
{
    return true // <== here
}
```

or would we end up with

```swift
if long
    + condition
{
    return true } // <== yuck
```

?

Reply:
Obviously(?) I prefer the former.

Reply:
Open and close breaks always guarantee paired consistency†, so if the user inserts a discretionary line break after the open delimiter then we force a line break before the close delimiter as well, and vice versa. So the former, desired outcome is what we'll get.

† The exception to this is for the close break before the right parenthesis/bracket of a function call/subscript expression, which we explicitly allow to remain on the previous line. (And this is controllable using a property on the break that can be adjusted on a case-by-case basis.)